### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False report with an 'as' assertion

### ISSUE THREAD

**Issue by FleetAdmiralJakob** [typescript-eslint/typescript-eslint#10257]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.2&fileType=.tsx&code=MYewdgzgLgBNCGUCmAuGAiJAPADgSwCckATdGAHwxyTGLzAHMyBeKmux9AbgCgeAzAK5hgUPOBgBbePQAUAShgBvHjBhEoggmGWq1MKAE9qaAAYASJQmQBfAEpIAjoKTRTMeBBgWlR6iH44KEQkeycXNz0bXhsgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0YANrhsOOdA7RImgLoaAviGNA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const state: "expired" | "pending" = "pending";

function main() {
  return {
    type: `${state}Request` as `${typeof state}Request`
  };
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

The code should be fine as the type assertion changes the type from `string` to `"expiredRequest" | "pendingRequest"`

### Actual Result

There was an error 5th line saying:
```
This assertion is unnecessary since it does not change the type of the expression. 5:11 - 5:56
```

**Comment by Josh-Cena** [comment]
This is because `as` is a valid place for contextual type inference. Just change `as` to `satisfies` and you will preserve the function return type.

**Comment by FleetAdmiralJakob** [comment]
But then it's not an unnecessary type assertion either. I mean you can use the type assertion `satisfies` is just better.

I mean the error is a little bit misleading as the `quick fix` suggests removing the type assertion not refactoring it to `satisfies`. In my case the type assertion / or the satisfies (either of them) is necessary, not optional.

Either the error should be removed or the `quick fix` updated I suggest.

**Comment by bradzacher** [comment]
Alternatively you can use `as const` for this.

**Comment by kirkwaiblinger** [comment]
Noting that this is closely related to #8721

**Comment by JoshuaKGoldberg** [comment]
Reopening since we reverted #10523.

**Comment by nicolo-ribaudo** [comment]
I have a similar bug, and I'm not sure if it's the same as this one or a different one because, differently from the one in this PR, changing `as` to `satisfies` as suggested above (https://github.com/typescript-eslint/typescript-eslint/issues/10257#issuecomment-2451118196) changes the types of the code

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABMMAKA7gLkQIgIY6IA%2BuARjgJSIDeAUIohAgM5SIAOeUUApgE5IAvIgAGAEmroAviMR5moiVACe7HnGCJpIgNy16iPjyggBHLrwF6p%2BgPS3EASU3K4IQzwC2cAG49EUAAW-jjgYDwQPMzMeHzKctH8sAiEeKS%2BPAA0AYEwCvx8cHzMtAA2xh7M2PiEJDjkiMIoqDUUekA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false 

```ts
function fn(w: "a" | "b") {
  const pattern = `${w}` as `${typeof w}`;

  return pattern;
}

// If you remove the "unnecessary assertion" above, this errors
let res: "a" | "b" = fn("a");
```
